python/ql/src/semmle/python/objects/ObjectAPI.qll

```diff
@@ -505,6 +525,20 @@ abstract class FunctionValue extends CallableValue {
     predicate isOverriddenMethod() {
         exists(Value f | f.overrides(this))
     }
+    
+    /** Whether `name` is a legal argument name for this function */
+    bindingset[name]
+    predicate isLegalArgumentName(string name) {
+        this.getScope().getAnArg().asName().getId() = name
+        or
+        this.getScope().getAKeywordOnlyArg().getId() = name
+        or
+        this.getScope().hasKwArg()
+    }
+    
+    predicate isAbstract() {
+        this.getARaisedType() = ClassValue::notImplementedError()
+    }
```

Comment:
I guess this is copied directly from Object, but I personally don't understand what is meant here. What does it mean that a function is abstract?

Does it have anything to do with [abstract base classes](https://docs.python.org/3/library/abc.html)?

Anyway, would be great if you could add some qldoc so it becomes crystal clear what this is about :blush:

Reply:
That is an excellent question. I have on the one hand a desire to say, well, what it means to be abstract is to raise a not implemented error! :p But on the other hand, I have no idea why "abstract" has something to do with non-implementation. It may indeed be related to abstract base classes. I think a lot of this has to do with whatever Mark's thoughts were at the time he authored this, but there's no documentation on the original so it's impossible to know without asking Mark. :\

Reply:
:smile: oh yeah, I know the pain :neutral_face: 

The reason I'm a bit stubborn about this, is that this is the perfect chance to get rid of old nasty code that we don't understand. If we first add it to `ObjectAPI.qll`, isn't going to be a pain to get rid off (since we can't know if a user/customer is using it).

I can see that it's only `Expressions/WrongNumberArgumentsInCall.ql` and `Expressions/WrongNameForArgumentInCall.ql` that uses `FunctionObject.isAbstract()`, to reduce the amount of FPs (I assume). There is a test-case illustrating this at https://github.com/Semmle/ql/blob/05b69a1c0faa794609e86160587c79ced0e941b9/python/ql/test/query-tests/Functions/overriding/test.py#L45-L59

The way I see it, `FunctionValue.getARaisedType` would be very cool to have, but I'm not so keen on the `FunctionValue.isAbstract()` -- could we maybe just move the logic to `Expressions/CallArgs.qll`, which is included by both of the above queries?

Reply:
Sounds reasonable. I can make it a predicate there.

Alternatively, since the definition of abstractness really is just.. "raises a `NotImplementedError`", we might just abandon the predicate entirely???

Reply:
That sounds reasonable to me! We could just inline `this.getARaisedType() = ClassValue::notImplementedError()`, and the not worry about naming :smile:

Reply:
It looks like this predicate is used in lots of queries so we can't remove it. I've added a semi-boring explanation??

Reply:
I don't think so. Notice that `ClassObject` also defines a predicate `isAbstract`.

```
$ git checkout master
$ cd ql/python/ql/src
$ git grep -E 'isAbstract\W' 
Classes/ClassAttributes.qll:      this.isAbstract()
Classes/MethodCallOrder.qll:    not self.isAbstract()
Classes/MissingCallToInit.ql:    not self.isAbstract()
Expressions/WrongNameForArgumentInCall.ql:not func.isAbstract() and
Expressions/WrongNumberArgumentsInCall.ql:not func.isAbstract() and
semmle/python/types/ClassObject.qll:    predicate isAbstract() {
semmle/python/types/FunctionObject.qll:    predicate isAbstract() {
```

it's only the 2 queries under `Expressions/` that use the `FunctionObject.isAbstract` predicate.

Reply:
@BekaValentine it looks like you fixed this?

Reply:
Yes, it's been moved to a stand alone predicate in `CallArgs.qll`